### ISSUE TITLE
Add an intermediary loader interface

### ISSUE THREAD

**Issue by ssttevee** [99designs/gqlgen#543]
Consider the following schema.

```
type Foo {}
type Bar {}
type Object {
  id ID!
  foo Foo
  bar Bar
}
```

I am trying to save on database queries by passing keys to type resolvers and then having each field resolver call `loader.Load`, except for the `id` field, instead of calling `loader.Load` once and passing the full object to the resolver like the [sample](https://gqlgen.com/reference/dataloaders/) in the documentation. Because of caching on the client side, most queries ask for only the `id` and no other fields, then asks for the full object using another query if it's not found in the cache.

```
func (r *ObjectResolver) ID(ctx context.Context, key DBKey) (ID, error) {
    return key.ID, nil
}

func (r *ObjectResolver) Foo(ctx context.Context, key DBKey) (*Foo, error) {
    obj, err := ctx.Value(objectLoaderKey).(*ObjectLoader).Load(key.ID)
    if err != nil {
        return nil, err
    }

    return obj.Foo, nil
}

func (r *ObjectResolver) Bar(ctx context.Context, key DBKey) (*Bar, error) {
    obj, err := ctx.Value(objectLoaderKey).(*ObjectLoader).Load(key.ID)
    if err != nil {
        return nil, err
    }

    return obj.Bar, nil
}
```

By doing this, the `id` field is effectively free since it is retrieved with the parent object. However, with larger types and queries, I'm finding that a non-trival amount of time is lost to context switching while calling `loader.Load` a large number of times. I can get better performance by using an intermediary model that has a method for every field of the real model other than `id`.

```
type InterimObject struct {
    key  DBKey
    once sync.Once
    obj  *Object
    err  error
    ctx context.Context
}

func (o *InterimObject) ID() ID {
    return o.key.ID
}

func (o *InterimObject) load() {
    o.obj, o.err = o.ctx.Value(objectLoaderKey).(*ObjectLoader).Load(o.key.ID)
}

func (o *InterimObject) object() (*Object, error) {
    obj.once.Do(o.load)
    return o.obj, o.err
}

func (o *InterimObject) Foo() (*Foo, error) {
    obj, err := o.object()
    if err != nil {
        return nil, err
    }

    return obj.Foo, nil
}

func (o *InterimObject) Bar() (*Bar, error) {
    obj, err := o.object()
    if err != nil {
        return nil, err
    }

    return obj.Bar, nil
}
```

This cuts the time spent calling `loader.Load`, but requires too much boilerplate code and errors are repeated once for each field.

I would like to propose an interface (that is not necessarily an interface) used to resolve objects prior to being passed resolvers when the field or method is not found, allowing the binding of the methods and fields of both the resolved object and the underlying struct.

Here's a hypothetical example:

```
type InterimObject struct {
    key  DBKey
}

func (o *InterimObject) ID() ID {
    return o.key.ID
}

func (o *InterimObject) Load(ctx context.Context) (*Object, error) {
    return ctx.Value(objectLoaderKey).(*ObjectLoader).Load(o.key.ID)
}
```

Given the above schema and model, this code may be generated:

```
func (ec *executionContext) _Object(ctx context.Context, sel ast.SelectionSet, obj *InterimObject) graphql.Marshaler {
	fields := graphql.CollectFields(ctx, sel, objectImplementors)

	var invalid bool

	var resVal *Object
	var resErr error
	load := func() {
		resVal, resErr = obj.Load(ctx)
		if resErr != nil {
			ec.Error(ctx, resErr)
			invalid = true
		}
	}

	var once sync.Once
	resolveObject := func() (*Object, error) {
		once.Do(load)
		return resVal, resErr
	}

	var wg sync.WaitGroup
	out := graphql.NewOrderedMap(len(fields))
	for i, field := range fields {
		out.Keys[i] = field.Alias

		switch field.Name {
		case "__typename":
			out.Values[i] = graphql.MarshalString("Object")
		case "id":
			wg.Add(1)
			go func(i int, field graphql.CollectedField) {
				out.Values[i] = ec._Object_id(ctx, field, obj)
				if out.Values[i] == graphql.Null {
					invalid = true
				}
				wg.Done()
			}(i, field)
		case "foo":
			wg.Add(1)
			go func(i int, field graphql.CollectedField) {
				defer wg.Done()
				res, err := resolveObject()
				if err != nil {
					return
				}
				out.Values[i] = ec._Object_foo(ctx, field, res)
			}(i, field)
		case "bar":
			wg.Add(1)
			go func(i int, field graphql.CollectedField) {
				defer wg.Done()
				res, err := resolveObject()
				if err != nil {
					return
				}
				out.Values[i] = ec._Object_bar(ctx, field, res)
			}(i, field)
		default:
			panic("unknown field " + strconv.Quote(field.Name))
		}
	}
	wg.Wait()
	if invalid {
		return graphql.Null
	}
	return out
}

func (ec *executionContext) _Object_id(ctx context.Context, field graphql.CollectedField, obj *InterimObject) graphql.Marshaler {
	...
}

func (ec *executionContext) _Object_foo(ctx context.Context, field graphql.CollectedField, obj *Object) graphql.Marshaler {
	...
}

func (ec *executionContext) _Object_bar(ctx context.Context, field graphql.CollectedField, obj *Object) graphql.Marshaler {
	...
}
```

**Comment by vektah** [comment]
This sounds simple but its a little tricky. How does gqlgen know to treat ID as special? What if the object came from a search by date? `query { find(after: "2018") { id } }` the above code would return 0s for everything.

Additionally there are a bunch of performance considerations:
 - this adds goroutines around all the primative fields which would make #518 much worse.
 - we want to make type safe batch loading easier, this is adding more boilerplate.

Rather than wrapping objects like this, I would reccomend calling CollectFields and checking in your resolver if only the id is present. After #533 lands it might be something like:
```go
func (r *ObjectResolver) Foo(ctx context.Context, key DBKey) (*Foo, error) {
    if graphql.CollectedFieldNames(ctx) == []string{"id"} {
        return Foo{ ID: key.ID }, nil
    }
    obj, err := ctx.Value(objectLoaderKey).(*ObjectLoader).Load(key.ID)
    if err != nil {
        return nil, err
    }

    return obj.Foo, nil
}
```

**Comment by ssttevee** [comment]
> How does gqlgen know to treat ID as special?

In the example I give, ID is treated differently because is part of the interim object. Gqlgen could first bind fields to the interim object, then check for a loader if there are missing fields before generating a field resolver.

> * this adds goroutines around all the primative fields

I don't think I actually added more goroutines. I just added the `sync.Once` related code to the goroutine that was generated from the schema using the latest commit on master. However, I can see how that would impede on the solution of #518.

> Rather than wrapping objects like this...

This seems like a sound solution. But the current `graphql.CollectFields` looks like it could be pretty expensive. If it reflects how `graphql.CollectedFieldNames` might be implemented, I'm not sure that it would be a good idea to use it on every query.
